### PR TITLE
add PG to SAM/BAM/CRAM files

### LINKED ISSUE #1527: SAM/BAM/CRAM writing tools need to add PR tag (optionally)
this is required for production, to track provenance

### PR DESCRIPTION
**Pull request by akiezun**
fixes #1527

added a bunch of small doc to GATKTool and made some methods final
@droazen can you review?

### CODE REVIEW

**Comment by akiezun** [comment]
@lbergelson cab you have a look? David is full HC

File: src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineProgram.java [hunk 1]
```diff
@@ -214,27 +217,41 @@ public String getStandardUsagePreamble() {
         return standardUsagePreamble;
     }
 
-    public CommandLineParser getCommandLineParser() {
+    /**
+     * Returns the command line parser of null if there is none.
+     */
+    public final CommandLineParser getCommandLineParser() {
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineProgram.java]: We could probably remove this method.  We don't use it at all.

File: src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java [hunk 2]
```diff
@@ -49,6 +50,9 @@
     @Argument(fullName="createOutputBamMD5", shortName="createOutputBamMD5", doc = "If true, create a MD5 digest for any BAM/SAM/CRAM file created", optional=true)
     public boolean createOutputBamMD5 = false;
 
+    @Argument(fullName="addOutputSAMProgramRecord", shortName="addOutputSAMProgramRecord", doc = "If true, adds a PG tag to created SAM/BAM/CRAM files.", optional=true)
+    public boolean addOutputSAMProgramRecord = false;
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java]: should this be on by default?

File: src/test/java/org/broadinstitute/hellbender/tools/PrintReadsIntegrationTest.java [hunk 3]
```diff
@@ -151,4 +152,29 @@ public void testReadThatConsumesNoReferenceBases() throws IOException {
         //Make sure we print the read, ie not lose it.
         SamAssertionUtils.assertSamsEqual(outFile, zeroRefBasesReadBam);
     }
+
+    @Test
```
Comment (lbergelson) [inline src/test/java/org/broadinstitute/hellbender/tools/PrintReadsIntegrationTest.java]: Should this be called testNoConflict**PG** ?

File: src/test/java/org/broadinstitute/hellbender/tools/PrintReadsIntegrationTest.java [hunk 4]
```diff
@@ -151,4 +152,29 @@ public void testReadThatConsumesNoReferenceBases() throws IOException {
...
+    public void testNoConflictRG() throws IOException {
+        final File inFile = new File(TEST_DATA_DIR, "print_reads_withPG.sam");
+        final File outFile = BaseTest.createTempFile("testNoConflictRG", ".sam");
+        final String[] args = new String[] {
+                "--input" , inFile.getAbsolutePath(),
+                "--addOutputSAMProgramRecord",
+                "--output", outFile.getAbsolutePath()
+        };
+        // Make sure no exception is thrown given an input containing a read that consumes no reference bases
```
Comment (lbergelson) [inline src/test/java/org/broadinstitute/hellbender/tools/PrintReadsIntegrationTest.java]: Does this file actually have a read that doesn't consume reference bases?  Which one is it?  It looks like they all have a section of Match except for the unmapped one.

**Comment by lbergelson** [comment]
@akiezun Looks good.  A few questions.  Merge when those are resolved.

File: src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineProgram.java (hunk 1 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineProgram.java]: done

File: src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java (hunk 2 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java]: i'm fine either way. Let's leave as off for ease of testing but it's easy to flip if the need arises.

File: src/test/java/org/broadinstitute/hellbender/tools/PrintReadsIntegrationTest.java (hunk 3 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/PrintReadsIntegrationTest.java]: yes, fixed

File: src/test/java/org/broadinstitute/hellbender/tools/PrintReadsIntegrationTest.java (hunk 4 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/PrintReadsIntegrationTest.java]: the comment was bogus. deleted now

**Pull request merged**